web_infrastructure/jenkins_plugin.py

```diff
@@ -0,0 +1,819 @@
+                        self.params['updates_url'],
+                        self.params['name'],
+                        self.params['version']))
+
+            if (
+                    self.params['updates_expiration'] == 0 or
+                    self.params['version'] not in [None, 'latest'] or
+                    md5sum_old is None):
+
+                # Download the plugin file directly
+                r = self._download_plugin(plugin_url)
+
+                # Write downloaded plugin into file if checksums don't match
+                if md5sum_old is None:
+                    # No previously installed plugin
+                    if not self.module.check_mode:
+                        self._write_file(plugin_file, r)
+
+                    changed = True
+                else:
+                    # Get data for the MD5
+                    data = r.read()
+
+                    # Make new checksum
+                    md5sum_new = hashlib.md5(data).hexdigest()
+
+                    # If the checksum is different from the currently installed
+                    # plugin, store the new plugin
+                    if md5sum_old != md5sum_new:
+                        if not self.module.check_mode:
+                            self._write_file(plugin_file, data)
+
+                        changed = True
+            else:
+                # Check for update from the updates JSON file
+                plugin_data = self._download_updates()
+
+                try:
+                    sha1_old = hashlib.sha1(open(plugin_file, 'rb').read())
+                except Exception:
+                    e = get_exception()
+                    self.module.fail_json(
+                        msg="Cannot calculate SHA1 of the old plugin.",
+                        details=e.message)
+
+                sha1sum_old = base64.b64encode(sha1_old.digest())
+
+                # If the latest version changed, download it
+                if sha1sum_old != plugin_data['sha1']:
+                    if not self.module.check_mode:
+                        r = self._download_plugin(plugin_url)
+                        self._write_file(plugin_file, r)
+
+                    changed = True
+
+        # Change file attributes if needed
+        if os.path.isfile(plugin_file):
+            params = {
+                'dest': plugin_file
+            }
+            params.update(self.params)
+            file_args = self.module.load_file_common_arguments(params)
+
+            if not self.module.check_mode:
+                # Not sure how to run this in the check mode
+                changed = self.module.set_fs_attributes_if_different(
+                    file_args, changed)
+            else:
+                # See the comment above
+                changed = True
+
+        return changed
+
+    def _download_updates(self):
+        updates_filename = 'update-center.json'
+        updates_file = "%s/%s" % (os.path.expanduser('~'), updates_filename)
```

Comment:
The second option looks better to me. Less required parameters is better. Let me fix it.